drivers/phoenixcontact_modbus.c

```diff
@@ -0,0 +1,200 @@
+{
+	upsdebugx(2, "upsdrv_updateinfo");
+    
+    uint16_t tab_reg[64];
+
+    mrir(ctx, 29697, 3, tab_reg);
+
+    status_init(); 
+
+    if (tab_reg[0])
+        status_set("OB");
+    else
+        status_set("OL");
+
+    if (tab_reg[2]) {
+        status_set("CHRG");
+    }
+
+    if (tab_reg[1]) {
+        status_set("LB"); //LB is actually called "shutdown event" on this ups
+    }
+
+    mrir(ctx, 29745, 1, tab_reg);
+    dstate_setinfo("output.voltage","%d",(int)(tab_reg[0]/1000));
+
+    mrir(ctx,29749,5,tab_reg);
+    dstate_setinfo("battery.charge","%d",tab_reg[0]);
+    //dstate_setinfo("battery.runtime",tab_reg[1]*60);
+
+    mrir(ctx,29792,10,tab_reg);
+    dstate_setinfo("battery.voltage","%f",(float)(tab_reg[0])/1000.0);
+    dstate_setinfo("battery.temperature","%d",tab_reg[1]-273);
+    dstate_setinfo("battery.runtime","%d",tab_reg[3]);
+    dstate_setinfo("battery.capacity","%d",tab_reg[8]*10);
+    dstate_setinfo("output.current","%f",(float)(tab_reg[6])/1000.0);
+
+    //ALARMS
+    mrir(ctx,29840,1,tab_reg);
+    alarm_init();
+    if (CHECK_BIT(tab_reg[0], 4) && CHECK_BIT(tab_reg[0], 5)) alarm_set("End of life (Resistance)");
+    if (CHECK_BIT(tab_reg[0], 6)) alarm_set("End of life (Time)");
+    if (CHECK_BIT(tab_reg[0], 7)) alarm_set("End of life (Voltage)");
+    if (CHECK_BIT(tab_reg[0], 9)) alarm_set("No Battery");
+    if (CHECK_BIT(tab_reg[0], 10)) alarm_set("Inconsistent technology");
+    if (CHECK_BIT(tab_reg[0], 11)) alarm_set("Overload Cutoff");
+    if (CHECK_BIT(tab_reg[0], 12)) alarm_set("Low Battery (Voltage)");
+    if (CHECK_BIT(tab_reg[0], 13)) alarm_set("Low Battery (Charge)");
+    if (CHECK_BIT(tab_reg[0], 14)) alarm_set("Low Battery (Time)");
+    if (CHECK_BIT(tab_reg[0], 16)) alarm_set("Low Battery (Service)");
+    alarm_commit();
+
+    status_commit();
+    if (errcount == 0)
+        dstate_dataok();
+
+}
+
+void upsdrv_shutdown(void)
+{
+	fatalx(EXIT_FAILURE, "shutdown not supported");
+}
+
+/*
+static int instcmd(const char *cmdname, const char *extra)
+{
+	if (!strcasecmp(cmdname, "test.battery.stop")) {
+		ser_send_buf(upsfd, ...);
+		return STAT_INSTCMD_HANDLED;
+	}
+
+	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
+	return STAT_INSTCMD_UNKNOWN;
+}
+*/
+
+/*
+static int setvar(const char *varname, const char *val)
+{
+	if (!strcasecmp(varname, "ups.test.interval")) {
+		ser_send_buf(upsfd, ...);
+		return STAT_SET_HANDLED;
+	}
+
+	upslogx(LOG_NOTICE, "setvar: unknown variable [%s]", varname);
+	return STAT_SET_UNKNOWN;
+}
+*/
+
+void upsdrv_help(void)
+{
+}
+
+/* list flags and values that you want to receive via -x */
+void upsdrv_makevartable(void)
+{
+	/* allow '-x xyzzy' */
+	/* addvar(VAR_FLAG, "xyzzy", "Enable xyzzy mode"); */
+
+	/* allow '-x foo=<some value>' */
+	/* addvar(VAR_VALUE, "foo", "Override foo setting"); */
+}
+
+void upsdrv_initups(void)
+{
+    int r;
+	upsdebugx(2, "upsdrv_initups");
+
+    ctx = modbus_new_rtu(device_path, 115200, 'E', 8, 1);
```

Comment:
BTW, should it (and other bits of the setting) be configurable? Either for this device in particular (does it accept different settings) or for generalization of the driver at a later stage?

Anyhow, this bit to interact more with NUT settings/config engine can be added at a later stage, as another PR.